Add a Settings page to configure people in workspace

mountiny opened the pull request:
### Details

This PR adds a workspace configuration page where admins can remove people from their workspaces or use modal to invite new members.

cc @marcaaron @MitchExpensify @Jag96 @madmax330 

### Fixed Issues
<!-- Please replace GH_LINK with the link to the GitHub issue this Pull Request is fixing -->
Fixes https://github.com/Expensify/Expensify/issues/166468

### Tests
1. Create a free workspace policy via the global create menu
1. Navigate to the settings screen and verify this workspace is visible
1. Tap the workspace
1. Tap the "People" option on the next screen
1. If there is only you, the admin, in the workspace, use the green `Invite` button which will bring in a modal from a side. Enter an email of an account you want to add to the workspace.
1. Once you added the number of accounts to the workspace you would wish, select some using the checkboxes on the left side. As soon as you select one checkbox, the `Remove` button should turn bright red. Click on the `Remove` button.
1. A confirmation modal looking like this should show up. Click `Remove` again to confirm the removal.
![image](https://user-images.githubusercontent.com/36083550/123350880-ce97f880-d553-11eb-85e4-cf9c8b8e93bf.png)
1. Make sure the person selected has been removed from the list and it does not show up even after page refresh.
1. Now, try selecting the first uppermost checkbox and make sure that all the checkboxes has been checked.
1. Go on and try to remove all of the members of the workspace.
1. Your account, the admin, should stay unremoved after this action. That is the correct behaviour, you cannot remove admin of workspace through this page.

### QA Steps
<!--- 
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image 
--->

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
![2021-06-30_10-43-36](https://user-images.githubusercontent.com/32969087/124028740-107bd100-d990-11eb-8edf-db478fa31ae8.png)

#### Mobile Web
![2021-06-30_10-48-56](https://user-images.githubusercontent.com/32969087/124029497-eaa2fc00-d990-11eb-8cd8-f431d8a5ace3.png)

#### Desktop
![2021-06-30_10-53-24](https://user-images.githubusercontent.com/32969087/124029985-6ac96180-d991-11eb-97b7-321bf72befa7.png)

#### iOS
![2021-06-30_10-41-48](https://user-images.githubusercontent.com/32969087/124028754-140f5800-d990-11eb-936c-17cd53aa0105.png)

#### Android
![2021-06-30_10-55-46](https://user-images.githubusercontent.com/32969087/124030273-bd0a8280-d991-11eb-9257-91a4e606f71e.png)

mountiny committed:
Add a new route

mountiny commented:
I will be making progress on this one over the weekend so I can ask clarifying questions on Monday.

mountiny committed:
Update languages

mountiny committed:
Revert merge changes

mountiny commented:
@marcaaron Sorry for polluting the PR with merged commits from Rajat's branch. Should I squash the commits to one or can I leave it like that?

@MitchExpensify Progress update:

I have shamelessly used great work from Rajat and Joe. I have added the People page, so far with title, tagline and the two buttons where only Invite button does what it should so far.

I will need to update the `src/styles/themes/default.js` with appropriate `buttonDangerBG` and `buttonDangerHoveredBG` and anything needed there as E.cash does not have red themed buttons yet. 

I have set up state property `selectedEmployees`, which will be used to keep track of the people checked to be potentially removed.

Tomorrow I will work on the table and `remove` functionality.

mountiny committed:
Align the buttons to left

mountiny commented:
@marcaaron I guess with the commit history, we can just leave it like this and once [Rajat's PR](https://github.com/Expensify/Expensify.cash/pull/3602) will be approved and merged into main, this diff will disappear.

mountiny committed:
Add button danger style option

mountiny committed:
Update Button component to have danger option

marcaaron commented:
@marcaaron Sorry for polluting the PR with merged commits from Rajat's branch. Should I squash the commits to one or can I leave it like that?

Squashing commits is fine as long as reviews have not yet started.

>@marcaaron I guess with the commit history, we can just leave it like this and once Rajat's PR will be approved and merged into main, this diff will disappear.

I would maybe suggest we take out the unmerged changes from Rajat's PR and set up this view as as standalone page for now. We can always tie everything together later and it would make the changes here easier to review (or we can wait until Rajat's changes are merged, but that seems less ideal).

mountiny committed:
Add the list of employees in the workspace

mountiny commented:
@marcaaron 
Update: I have used the `InvertedFlatList` to create the list of employees. The styling will definitely need some polishing to match the proposed design. 

Can you please give me a hint what is the best way for writing something like `@media` queries as we need to change the cell design slightly for mobile?

For now, I have included mockData for easier troubleshooting of the design and in case you do not have the correct `policyId` yet. But it works with `employeeList` too.

Next, I will work on adding the the Remove onyx functionality. The checkboxes on page should already work.  i will also look into taking the Rajat changes out.

@shawnborton I thought you might be the best one to ask regarding design, for the Admin badge, there is the little tick. Is that already somewhere in code or is it something new I should add. In that case, would you have an example of something similar? Thank you!

@marcaaron I wanted to ask for a clarification to this last bullet point of the doc:
> A badge that indicates the admin (hint: it’s only the current user for now and will be used from ONYXKEYS.SESSION)

I cannot see anything helpful to find out if the user is admin in `SESSION`. Is that something which is being implemented somewhere else? Or who should I show the admin badge at now? Thank you!

Thank you guys for any feedback on the code so far!

![image](https://user-images.githubusercontent.com/36083550/122839835-be341380-d2f0-11eb-91dc-c63a1888fcd6.png)

![image](https://user-images.githubusercontent.com/36083550/122840449-03a51080-d2f2-11eb-819f-c531b9c9aea0.png)

marcaaron commented:
>Update: I have used the InvertedFlatList to create the list of employees. The styling will definitely need some polishing to match the proposed design.

We probably can just use the regular `FlatList` that comes with React Native as we don't need to invert the content for this case (only for a chat and anything that has content from the bottom to top).

>Can you please give me a hint what is the best way for writing something like @media queries as we need to change the cell design slightly for mobile?

https://github.com/Expensify/Expensify.cash/blob/main/src/components/withWindowDimensions.js

>I cannot see anything helpful to find out if the user is admin in SESSION. Is that something which is being implemented somewhere else? Or who should I show the admin badge at now? Thank you!

We can access the current user's login via `session` key and then compare to the admin email. For example, here we check to see if the currently logged in user has a "phone login" so we can do something similar to that...

https://github.com/Expensify/Expensify.cash/blob/c1536a83ef628f27dcae5d0320bfba16d7bff58e/src/pages/ReimbursementAccount/ReimbursementAccountPage.js#L112

shawnborton commented:
@mountiny  the asset already exists as an .svg, it's probably named `checkmark.svg`

There is some other design feedback I have too about the overall style of the modal, but I know we have another contributor working on some of these styles for the Expensify Card stuff. @marcaaron what do you recommend we do here? Wait and do a follow up?

marcaaron commented:
Sounds fine as long as the focus is on what's inside the modal @shawnborton? The rest we can cover with Rajat since we should be removing the code from here so we can tackle the relevant changes in each. So just this area and not the overall style of the modal.

![2021-06-21_15-26-07](https://user-images.githubusercontent.com/32969087/122848349-10e7de00-d2a5-11eb-9290-9945f211d80d.png)

shawnborton commented:
Cool, that works for me!

mountiny commented:
@shawnborton Feel free to leave the design feedback here, I will tackle this first thing in a morning tomorrow and hopefully get close to finish!

shawnborton commented:
Things look good to me. But actually @michelle-thompson can you remind me, what do we need the checkmark for in the little `Admin` pill?

michelle-thompson commented:
I think we actually decided against that since this table will only show who's already been added to the workspace.

shawnborton commented:
Oh nice, that was easy. Mind adding an up-to-date mock here when you get a chance?

michelle-thompson commented:
Hmm, none of the above mockups have checkmarks in the pills though?

Here it is in case I'm missing something:
![image](https://user-images.githubusercontent.com/43361905/122991905-69c57e00-d35a-11eb-8e49-405c8defed41.png)

shawnborton commented:
Check out the linked issue though: https://github.com/Expensify/Expensify/issues/166468 You will see the checkmark there

mountiny committed:
Update admin badge style

mountiny committed:
Revert: Refactor the pages so the settings is a standalone page

mountiny committed:
Update the features after refactor

mountiny committed:
Remove Rajat's components

mountiny committed:
Use simply FlatList and add disabled danger button colour

mountiny commented:
@michelle-thompson Can I please ask, what is the hex code of the disabled red `remove` button? I have burnt too much time trying to do it with half opacity on the button I will just add a new `redDisabled` colour :)

mountiny commented:
@marcaaron I have refactored this PR so it does not use the commits and components from Rajat's PR. 

I have switched to using normal FlatList for the table. I got a follow-up question for the `admin` badge. In the `policy` key, there is a `role` key which says `admin` or `user`. Should we not use that? Or is it simply that whoever is not admin of the policy cannot even see this page, so we can assume the person who is signed in is the admin at the same time?

```
{
  employeeList: ["vit@expensify.com", "vit+2@expensifail.com", "vit+1@expensifail.com", "vit+1@expensify.com"],
  id: "F42142AB731F9782",
  name: "Expensify",
  role: "user",
  type: "team",
}
```

mountiny committed:
Add Policy member removal functionality

michelle-thompson commented:
@shawnborton might you know the specifics of the red disabled state?

mountiny commented:
UPDATE:
I have added the remove user from policy functionality. On the screenshot below, is the modal which shows up to confirm a removal. Any proposals to change the texts in the modal?
![image](https://user-images.githubusercontent.com/36083550/123182502-827f8200-d487-11eb-9248-98aa8c322da9.png)

@marcaaron  There is one UX issue I believe and that is that the admins cannot remove themselves from the policy using this API command. Therefore, checking the upper checkbox to tick all and then trying to remove it end up in failed request, thus not removing anyone. For better UX, we should not even give the user an option to try to do something they cannot do. 

So I propose to not display the checkbox for admin, as it would just be causing confusion for users in future.

Last thing before being ready for a review is the responsive design.  Gonna do it tomorrow and ask for a review.

mountiny committed:
Merge main into the branch to get up-to-date changes

mountiny committed:
Fix two merge conflict errors

mountiny committed:
Fix issue with undefined key

mountiny committed:
Fixing JS linting errors

mountiny committed:
Fix js lint

mountiny commented:
Additionally, I wanted ot ask if it expected behaviour to be able to invite emails which do not have Expensify account associated with them using the invite modal Joe has implemented  here https://github.com/Expensify/Expensify/issues/166040

I thought these lines in the response handler are there to handle the case in which user put in email which is not existing in the database. https://github.com/Expensify/Expensify.cash/blob/a8703e227b44bab8d6cf6685cd34077ef48b4bbf/src/libs/actions/Policy.js#L173-L175

mountiny committed:
Fix js lint 2

marcaaron commented:
>I have switched to using normal FlatList for the table. I got a follow-up question for the admin badge. In the policy key, there is a role key which says admin or user. Should we not use that? Or is it simply that whoever is not admin of the policy cannot even see this page, so we can assume the person who is signed in is the admin at the same time?

We can ignore the role for now and just use the current user email to determine the admin badge (the role will be used later as we add more options). For now, there is no ability to change a user's role and free policy should have one admin only.

>whoever is not admin of the policy cannot even see this page, so we can assume the person who is signed in is the admin at the same time

Correct, only admins will see the policy in their list in `/settings` according to this logic here

https://github.com/Expensify/Expensify.cash/blob/8f8d835171db5198a4c60167e5266eb5326fda51/src/pages/settings/InitialPage.js#L128-L130

>I have added the remove user from policy functionality. On the screenshot below, is the modal which shows up to confirm a removal. Any proposals to change the texts in the modal?

@shawnborton @michelle-thompson I didn't see any mockups for this, but did see it mentioned in the doc (in text at least). Do we need this confirmation step or if we can just remove these users? @mountiny if we're keeping the modal I'd say we can use the copy from Web-Expensify or asking marketing for feedback. Currently we have this...

>Are you sure you want to unshare the policy or remove the selected people?

I find it a bit confusing to refer to both a "workspace" and a "policy" and think the language on that could be aligned better for this flow.

>There is one UX issue I believe and that is that the admins cannot remove themselves from the policy using this API command. Therefore, checking the upper checkbox to tick all and then trying to remove it end up in failed request, thus not removing anyone. For better UX, we should not even give the user an option to try to do something they cannot do.

I agree this is strange (the same UX exists in OldDot), but I think we can improve this later. At a bare minimum, we just need to not throw any errors. Filtering out the current user before making the request is probably how I'd approach this.

>Additionally, I wanted ot ask if it expected behaviour to be able to invite emails which do not have Expensify account associated with them using the invite modal Joe has implemented here Expensify/Expensify#166040 I thought these lines in the response handler are there to handle the case in which user put in email which is not existing in the database.

We can test this by inviting a user to a policy when that user does not exist in the local database. The answer to why we catch this `402` code certainly rests [somewhere](https://github.com/Expensify/Web-Expensify/blob/a4a74f01b8080859f312ed0dd8dfc8748edb2b3b/api.php#L1469-L1478) in the API in Web-Expensify. Maybe a good question for `#engineering-chat` :)

mountiny committed:
Update the texts

mountiny committed:
Remove the admin email from list of members to remove to prevent error message

michelle-thompson commented:
@mountiny can you please change the `Remove` button to red? Also, @shawnborton I can't remember what the rule of thumb is for modal headers -- capitalized vs. regular. Might you know?

shawnborton commented:
Looks like we aren't consistent based on other examples:
![image](https://user-images.githubusercontent.com/2319350/123346067-4a6e5280-d50c-11eb-8579-fc7065b393ac.png)
![image](https://user-images.githubusercontent.com/2319350/123346093-5823d800-d50c-11eb-9a6c-221fb4a9150b.png)

Which do you prefer?

mountiny committed:
Add responsive styles and make remove button in modal red

mountiny commented:
@marcaaron 
> We can test this by inviting a user to a policy when that user does not exist in the local database

I have asked [in Slack](https://expensify.slack.com/archives/C03TQ48KC/p1624544719391900) about the ability of user to invite email which is not associated with any account and we will not do anything about it now. Only in case it will cause problems in future.

> I find it a bit confusing to refer to both a "workspace" and a "policy" and think the language on that could be aligned better for this flow.

I have updated the texts to refer to workspace so it is not so confusing. People who come to NewDot will not be familiar with the Policies most likely.

>  Filtering out the current user before making the request is probably how I'd approach this.

That is what I have done and additionally added a check to the `removeMembers` function to see if the members array is not empty (this would happen if the user selected only themselves being admin and filtering out their email leaves out empty array)

> Can you please change the Remove button to red?

@michelle-thompson Done!

I will write up some more about how to test this PR and ask for an initial review.

michelle-thompson commented:
I think it makes more sense to stick with capitalized.

mountiny committed:
Capitalize modal title

mountiny commented:
@michelle-thompson Made it capitalized, see the screenshot in PR description!

mountiny commented:
@marcaaron @TomatoToaster Should be ready for a first review 😊🙏

michelle-thompson commented:
Looks great to me!

mountiny committed:
Remove debugging line

marcaaron commented:
Hmm what is the copy here actually supposed to say? This sounds like a famous Bob Ross quote.

>In life you need colors. At home you have unlimited time. Maybe we got a few little happy bushes here.

![2021-06-25_06-50-04](https://user-images.githubusercontent.com/32969087/123459144-9ad3c780-d581-11eb-9b7c-71186b236686.png)

marcaaron commented:
Not sure if we need to address this right now but this feels a little squished to me

![2021-06-25_06-51-22](https://user-images.githubusercontent.com/32969087/123459298-cf478380-d581-11eb-99dd-edf2d3a55bb7.png)

shawnborton commented:
Hah, yeah that is totally Bob Ross placeholder text. 

I agree with the label feeling squished - we should handle this like the LHN where the email address part would truncate. Flexbox should be our friend here.

mountiny commented:
Haha, where should I ask for the actual copy?

shawnborton commented:
I'd say bring the screenshot into the #n5_5-free-plan room and see if anyone there is up for helping out. There is a `@marketingcontent` team in Slack so you could ping them and someone should be able to help with writing the copy. (I think that's how that works at least, I could be totally wrong)

marcaaron commented on src/libs/API.js:
```diff
@@ -956,6 +956,17 @@ function Policy_Create(parameters) {
+/**
+ * @param {Object} parameters
+ * @param {Array} [parameters.emailList]
+ * @param {String} [parameters.policyID]
```
I don't think `policyID` or `emailList` should be optional (that's what the `[]` mean in this JSDoc).

Also we should use `requireParameters()` [like we do for this command in Web-Expensify](https://github.com/Expensify/Web-Expensify/blob/c0a9aba689e24c179bb403f1fcc59e452b926198/site/lib/api.js#L743)

mountiny replied:
Ah, thanks for mentioning that, I thought it means something else in this case. Updated it!

mountiny commented:
Great @shawnborton, gonna do it and blame you for taking people's time if it is wrong 😄

marcaaron commented on src/libs/Navigation/AppNavigator/AuthScreens.js:
```diff
@@ -225,6 +226,15 @@ class AuthScreens extends React.Component {
+                    }}
+                    component={WorkspacePeopleModalStackNavigator}
+                    listeners={modalScreenListeners}
+                />
```
This is probably annoying (since I asked you to basically do the opposite), but since we have merged Rajat's PR and the changes are ready now I think we can just add this screen here instead of create the `WorkspacePeopleModalStackNavigator`.

https://github.com/Expensify/Expensify.cash/blob/b55729a4c27deb3ea97566aadcb2b6724a147387/src/libs/Navigation/AppNavigator/WorkspaceSettingsDrawerNavigator.js#L11-L17

Sorry about the confusion there. Mostly I was trying to communicate that it would be a good practice to avoid including unmerged changes in a PR (since it could potentially lead to a lot of changes for you and make reviewers have to look at things that are unrelated), but now that those changes **are merged** it does make sense to use them since we're introducing changes that we will need to immediately undo.

mountiny replied:
Of course, that is no problem, it was my mistake to include Rajat commits in here as well. I will merge `main` branch into this branch and update it so the page is part of the settings modal. Gonna do that later.

marcaaron commented on src/libs/actions/Policy.js:
```diff
@@ -91,6 +91,54 @@ function getPolicyList() {
+ */
+function removeMembers(members, policyID) {
+
+    // Check if the members array is not empty
```
It's pretty obvious what the check is doing. I'd suggest we do one of these:

1. Improve this comment so it explains "why" and not "what"
2. Remove this comment since it should be obvious to anyone looking at this why we would return early in this case.

mountiny replied:
I chose to update the comment as this is a consequence of deciding to filter out the admin email from member list in different file, therefore something which might not be clear to other developers when inspecting the code at first glance.

marcaaron commented on src/libs/actions/Policy.js:
```diff
@@ -91,6 +91,54 @@ function getPolicyList() {
+
+    // Make a shallow copy to preserve original data and remove the members
+    const policy = _.clone(allPolicies[key]);
+    policy.employeeList = policy.employeeList.filter(email => !members.includes(email));
```
Heads up, we default to using the `_` variety of array methods for consistency. 

In this case, however I'd recommend we use the [`_.without()` method](https://underscorejs.org/#without) instead of filter + includes.

marcaaron commented on src/libs/actions/Policy.js:
```diff
@@ -91,6 +91,54 @@ function getPolicyList() {
+                errorMessage += ` ${translateLocal('workspace.people.cannotRemovePolicyOwner')}`;
+            }
+
+            Growl.show(errorMessage, CONST.GROWL.ERROR, 5000);
```
NAB, do we expect any of these errors to appear for users? Seems like the 401 would happen if an admin could view this page but an admin should not be able to view the page. And we are already preventing from removing the policy owner by filtering them out. 

What we have here is _probably_ fine, but in the future we should make sure we are optimizing for the cases that we expect to encounter today and not the unexpected e.g. instead of catching these specific errors we could have shown just the generic message at the bare minimum or log to the console instead.

mountiny replied:
You are right, technically, we do not expect the two to show up since non-admins should not even be able to see the page and then we decided to filter out the policy owner from the list to remove if they selected themselves. 

I will update it so we only log them to the console and show the generic error message to user.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+        this.inviteUser = this.inviteUser.bind(this);
+        this.removeUsers = this.removeUsers.bind(this);
+        this.toggleUser = this.toggleUser.bind(this);
```
This does not need to be bound in the constructor since all usages are inside anonymous arrow functions whose scope is bound to the component already.

mountiny replied:
Very well spotted, thank you! Had to update how I call the functions too.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+    /**
+    * Open the modal to invite a user
+    */
```
This doc looks formatted incorrectly it should be 

```
/**
 * Open the modal to invite a user
 */
```

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+    /**
+    * Remove selected users from the workspace
+    */
```
Same here doc formatting

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+    /**
+    * Show the modal to confirm removal of the selected members
+    */
```
Formatting

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+    /**
+    * Hide the confirmation modal
+    */
```
Formatting

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    * Add or remove user from the selectedEmployees list
+    *
+    * @param {Sting} login
+    */
```
Formatting

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    removeUsers() {
+        // Remove the admin from the list
+        // eslint-disable-next-line max-len
+        removeMembers(this.state.selectedEmployees.filter(email => email !== this.props.session.email), this.props.route.params.policyID);
```
use `_.without()`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    */
+    removeUsers() {
+        // Remove the admin from the list
+        // eslint-disable-next-line max-len
```
I think it would be better to resolve this by creating a new variable for the members then pass that to `removeMembers()`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+        // eslint-disable-next-line max-len
+        removeMembers(this.state.selectedEmployees.filter(email => email !== this.props.session.email), this.props.route.params.policyID);
+        this.setState(prevState => ({
+            ...prevState,
```
There's no need to spread `prevState` in here react will know that you only want to modify the state values specified.

marcaaron replied:
And since we are not referencing the `prevState` we can call `this.setState({someProperty: value});`

mountiny replied:
I have been spreading the prevState unnecessarily for last 2 years 🙃 thanks for heads up!

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    * Show the modal to confirm removal of the selected members
+    */
+    askForConfirmationToRemove() {
+        this.setState(prevState => ({...prevState, isRemoveMembersConfirmModalVisible: true}));
```
```suggestion
        this.setState({isRemoveMembersConfirmModalVisible: true});
```

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    * Hide the confirmation modal
+    */
+    hideConfirmModal() {
+        this.setState(prevState => ({...prevState, isRemoveMembersConfirmModalVisible: false}));
```
`prevState`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    * @param {Sting} login
+    */
+    toggleUser(login) {
+        if (login === 'ALL') {
```
I think this might be clearer as a separate method `toggleAllUsers()` this way we don't need to introduce this constant or special significance for the `ALL` user and it will be very clear what will happen.

`toggleUser('ALL')` vs `toggleAllUsers() + toggleUser('some user')`

mountiny replied:
Ah missed this one. Yes, I agree, that will be clearer. I will refactor it.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+    toggleUser(login) {
+        if (login === 'ALL') {
+            this.setState(prevState => ({
+                ...prevState,
```
no spread prevState

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+        if (login === 'ALL') {
+            this.setState(prevState => ({
+                ...prevState,
+                selectedEmployees: prevState.selectedEmployees.length !== this.props.policy.employeeList.length
```
I'm unsure why we should compare the lengths of state and props here to determine which way we should toggle the users. It might be better to look at the state of this checkbox and then update the state based on that.

e.g. 

```js
this.setState(prevState => ({
    shouldSelectAllEmployees: !prevState.shouldSelectAllEmployees,
    selectedEmployees: !prevState.shouldSelectAllEmployees ? this.props.policy.employeeList : [],
}));
```

mountiny replied:
So the reason was that the `all` header checkbox state is not dependant on `this.state. shouldSelectAllEmployees` but on `this.state.selectedEmployees.length === this.props.policy.employeeList.length`.

I think I will actually drop the use of `shouldSelectAllEmployees` as it is making it unnecessarily complex as we would need to check on each toggle if all members have been selected and `shouldSelectAllEmployees` should be set to `true`.

marcaaron replied:
That seems fine to me, if we fix up the `toggleUser()` method as well.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+        }
+
+        this.setState(prevState => ({
+            ...prevState,
```
no spread prevState

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+
+        this.setState(prevState => ({
+            ...prevState,
+            selectedEmployees: prevState.selectedEmployees.includes(login)
```
`_.contains`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+        this.setState(prevState => ({
+            ...prevState,
+            selectedEmployees: prevState.selectedEmployees.includes(login)
+                ? prevState.selectedEmployees.filter(item => item !== login)
```
`_.filter`

mountiny replied:
I have used `_.without` here as we know the login is included, it will be a bit shorter than `_.filter`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+            ...prevState,
+            selectedEmployees: prevState.selectedEmployees.includes(login)
+                ? prevState.selectedEmployees.filter(item => item !== login)
+                : [...prevState.selectedEmployees, login],
```
It might be a little simpler to just do something like 

```js
selectedEmployees: _.uniq([...prevState.selectedEmployees, login]);
```

mountiny replied:
I think this would not work as we need to. `_.uniq` would always keep one instance of `login` in the array so we could not toggle the member `off`.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                            </View>
+                        </View>
+                    ) : (
+                        <Fragment>
```
NAB/FYI can also use the `<></>` syntax for fragment if we do not need them keyed

mountiny replied:
Perfect, was not sure if it is a best practise here. Gonna use it from now on.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+            <View style={[styles.peopleRow]}>
+                <View style={[styles.peopleRowCell, styles.peopleCheckbox]}>
+                    <Checkbox
+                        isChecked={this.state.selectedEmployees.includes(item.login)}
```
`_.contains`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+            <View style={[styles.peopleRow, styles.peopleHeaderRow]}>
+                <View style={[styles.peopleRowCell, styles.peopleCheckbox]}>
+                    <Checkbox
+                        isChecked={this.state.selectedEmployees.length === this.props.policy.employeeList.length}
```
Let's use something like `this.state.shouldSelectAllEmployees`? While the employeeList being equal in length to the selectedEmployees is the end result we are looking for and we _could_ use it to determine which way to toggle we'll need to do this same check every time.

mountiny replied:
I was thinking about this, but then I think it would end up in quite clumsy `toggleUser` function.

Let me know what you think and what should we choose. The function would have to look like this:

```
toggleUser(login) {
        if (login === 'ALL') {
            this.setState(prevState => ({
                shouldSelectAllEmployees: !prevState.shouldSelectAllEmployees,
                selectedEmployees: this.props.policy.employeeList.length !== prevState.selectedEmployees.length 
                    ? this.props.policy.employeeList 
                    : [],
            }));
            return;
        }

        this.setState(prevState => ({
            shouldSelectAllEmployees: ((!_.contains(prevState.selectedEmployees, login) && prevState.selectedEmployees.length === this.props.policy.employeeList.length - 1)
                ? true
                : false,
            selectedEmployees: _.contains(prevState.selectedEmployees, login)
                ? _.without(prevState.selectedEmployees, login)
                : [...prevState.selectedEmployees, login],
        }));
    }
```

The condition there is quite hard to read. We need to check if the `login` is in `prevState` and also if it is the last one 'unchecked'. We could write it in a longer more readable way but I am not sure if the advantages of having the `shouldSelectAllEmployees` state key are outweighing the worse readability. 

Let me know what you think!

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+            <ScreenWrapper>
+                <HeaderWithCloseButton
+                    title={this.props.translate('common.people')}
+                    onCloseButtonPress={() => Navigation.dismissModal(true)}
```
Why call `Navigation.dismissModal()` with a `true` argument?

mountiny replied:
My mistake

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                        <View style={styles.buttonRow}>
+                            <Button
+                                success
+                                style={[]}
```
Do we need this empty array passed to `style`?

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                this.props.policy.employeeList ? (
+                                    <FlatList
+                                        ListHeaderComponent={this.renderHeader()}
+                                        ref={flatListRef}
```
This seems incorrect as the ref is coming from `ReportScrollManager`

mountiny replied:
Removed!

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                        ref={flatListRef}
+                                        renderItem={this.renderItem}
+                                        contentContainerStyle={[styles.w100]}
+                                        initialRowHeight={32}
```
This seems unnecessary

mountiny replied:
The container style needs to be `width: 100%` otherwise the columns do not align well.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                            ? this.props.policy.employeeList.map(email => this.props.personalDetails[email] || false)
+                                            : []}
+                                        maxToRenderPerBatch={1}
+                                        windowSize={15}
```
`maxToRenderPerBatch` and `windowSize` seem like they can be removed for now.

mountiny replied:
Removed!

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                            : []}
+                                        maxToRenderPerBatch={1}
+                                        windowSize={15}
+                                        removeClippedSubviews={this.props.shouldRemoveClippedSubviews}
```
Doesn't seem like this prop exists. Also, should we `removeClippedSubviews` here or is it fine to "do nothing" for now? I think maybe it's fine.

mountiny replied:
Good point, I missed this.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                        data={(this.props.policy.employeeList && this.props.policy.employeeList.length !== 0)
+                                            // eslint-disable-next-line max-len
+                                            ? this.props.policy.employeeList.map(email => this.props.personalDetails[email] || false)
+                                            : []}
```
Rather than check the `employeeList` here it might be better to just initialize this in the beginning of `render()` with a new variable `data` in the correct shape the list wants.

mountiny replied:
That will be better for readability, good point!

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                        // eslint-disable-next-line max-len
+                                        data={(this.props.policy.employeeList && this.props.policy.employeeList.length !== 0)
+                                            // eslint-disable-next-line max-len
+                                            ? this.props.policy.employeeList.map(email => this.props.personalDetails[email] || false)
```
I think we will have an array with values like this ?

```js
['someone@exists.com', false, 'another@inpersonaldetails.com'];
```

What can we do for the people who were not found? Should we just filter them out? Show some default?

mountiny replied:
I have filtered them out for now. However, if users signs up using phone number, they might not have `email`, right? 

Should we show their mobile number instead?

marcaaron replied:
Ah, the answer is no. `email` is kind of a misleading/unfortunate variable name in this case. 

Even exclusive phone users have an `email` and it looks something like:

`+15141231234@expensify.sms`

mountiny replied:
Thank you! Great to know.

Do you reckon we should filter out these sm`expensify.sms` emails?

marcaaron replied:
I can't of a reason why we would. If it's possible to invite someone with a phone number to a workspace then we would show them in this list..

mountiny replied:
Should we then consider changing the header name of the column to be `Email or phone` same as the Invite modal has in placeholder of the input?

marcaaron replied:
It might be a good follow up issue, but I don't think we need to block on it.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,328 @@
+                                    <View style={[styles.peopleRow]}>
+                                        <View style={[styles.peopleRowCell, styles.flex2]}>
+                                            <Text>
+                                                Unfortunately, this workspace has no members.
```
1. I think maybe it's not possible since we should always have at least one member
2. If not, did we get this copy approved or where did it come from?
3. Why so serious?? 😂

mountiny replied:
1. You are right, I made this edge case without thinking about it too much.
2. Nope, that was just a placeholder.
3. 🥺 coz I dont want to mess up in my new job.

marcaaron reviewed (changes requested):
Overall looks pretty solid. Very nice work! But I do have some comments.

marcaaron commented:
In addition to the review comments there are conflicts that need to be fixed before this can be merged. Thanks!

mountiny commented:
It was decided [here](https://expensify.slack.com/archives/C020EPP9B9A/p1624640638334600) by the `@marketing-content` team that we will remove the tagline at all and keep it blank.

mountiny commented:
@shawnborton Do we want to keep the space growing for those rows where no Admin badge is placed like on the screen? Or cut it off for the rest who have no badges too?

![image](https://user-images.githubusercontent.com/36083550/123479067-4455a000-d5f8-11eb-9d59-02dca5263df4.png)

shawnborton commented:
I think what you have in your screenshot is perfect.

mountiny committed:
Improving the styles on smaller screen

mountiny committed:
Update the responsive styles and remove the tagline

mountiny commented on src/languages/en.js:
```diff
@@ -422,6 +425,15 @@ export default {
+            cannotRemovePolicyOwner: 'Can’t remove the admin of this workspace.',
+            removeMembersPrompt: 'Are you sure you want to remove the selected people from your workspace?',
+            removeMembersTitle: 'Remove Members',
+        },
```
Yes, we should, I wanted to leave that for end as I did not know how many new keys I will need to add to the `en.js.`

mountiny replied:
@marcaaron Should I tag someone Spanish speaking here to add the appropriate translations or what is the common process for these? I have pushed commit which adds the keys to `es.js`.

mountiny replied:
Thank you!

mountiny committed:
Add translation keys to es.js

mountiny committed:
Make the remove request parameters required

mountiny committed:
Update one comment

mountiny commented on src/components/Button.js:
```diff
@@ -96,8 +101,11 @@ const Button = (props) => {
+                        (props.isDisabled && props.danger) ? styles.buttonDangerDisabled : undefined,
+                        (props.isDisabled && !props.danger) ? styles.buttonDisable : undefined,
                         (props.success && hovered) ? styles.buttonSuccessHovered : undefined,
+                        (props.danger && hovered) ? styles.buttonDangerHovered : undefined,
```
Taking note of this and will create a refactoring issue after this PR is done 👍

mountiny committed:
Merge main into vit-settingsPageToConfigureWorkspace

mountiny committed:
Revert debugging changes and update remove request call error handling

mountiny committed:
Remove unnecessary funtion bindings

mountiny committed:
Fix JSDoc formatting

mountiny committed:
Refactor the toggleUser function

mountiny committed:
Remove fragment and use underscore contains

mountiny committed:
Remove unnecessary props for FlatList

mountiny committed:
Update responsive design breakpoint to work well in the modal

mountiny commented:
@marcaaron I have worked in the requested changes. I think there is just couple of things unresolved now in the comments above.

One thing we need to resolve is  that the modal sidebar in `WorkspaceSidebar.js` does not have `props.route` right now since it is not a page. We need the `policyID` to get correct link on the 'People' menu item. What do you suggest? Should we try to tunnel it down in props from `AuthScreen.js`? there should be some better way.

For now, I have commented it out as we need to resolve it before merging.

https://github.com/Expensify/Expensify.cash/blob/00a64506a8831b18fc43a35f24ab02b2f310c8d0/src/pages/workspace/WorkspaceSidebar.js#L37-L45

mountiny committed:
Fix js linting errors

marcaaron commented:
>One thing we need to resolve is that the modal sidebar in WorkspaceSidebar.js does not have props.route right now since it is not a page. We need the policyID to get correct link on the 'People' menu item. What do you suggest? Should we try to tunnel it down in props from AuthScreen.js? there should be some better way.

As mentioned elsewhere, we should hook this view up to the workspace drawer navigator and not make it a standalone page. Then we can access the policyID in the route.

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,327 @@
+        let data = [];
+        if (this.props.policy.employeeList && this.props.policy.employeeList.length !== 0) {
+            data = _.filter(this.props.policy.employeeList, email => this.props.personalDetails[email])
+                .map(email => this.props.personalDetails[email]);
```
There are a few ways to improve this none of them major but pointing them out anyways in case it helps..

1. `_.filter()` should always return at least an empty array so the initialization step with `let` is unnecessary
2. If we chain `.map()` after `_.filter` technically we are using the native array method and not underscore. It's more common to see use of [`_.chain()`](https://underscorejs.org/#chain) if you need to combine array methods
3. We can use `lodashGet()` to make sure we access the value or a default of an object
4. `_.filter` _should_ remove any undefined values by default (I think)

With all of that we'd end up with something like

```js
const data = _.chain(lodashGet(this.props, 'policy.employeeList', []))
    .map(email => this.props.personalDetails[email])
    .filter()   
    .value();
```

mountiny replied:
Great job! I will go ahead and get more familiar with the `underscore` functions since I would have never thought of that `chain()` function. 

Ad 1. I had to initialize it since I have had the conditional statement there so I could not have initialized the `data` variable inside of the if block and then use it later because of incorrect scope.

marcaaron commented:
I think I may have misread your comment above.. but `WorkspaceSidebar.js` can access the route already...

https://github.com/Expensify/Expensify.cash/blob/1a3344218c0f6ac500c340e7520fdc52cd0cbf47/src/pages/workspace/WorkspaceSidebar.js#L51

mountiny committed:
Fix grammar mistake

mountiny committed:
Refactor user toggling logic

mountiny committed:
Change the Confirm modal button props

mountiny commented on src/components/ConfirmModal.js:
```diff
@@ -62,7 +66,8 @@ const ConfirmModal = props => (
 
             <Button
-                success
+                success={!props.danger}
```
You are right, it will be better not inverted. The styles of `danger` version are overwriting the success so now we can use it with just `success` but as you mentioned before, we should refactor the `Button` component as a follow up.

mountiny committed:
Remove placeholder ModalStackNavigator

mountiny committed:
Update Remove members error handling and remove unnecessary translations

mountiny committed:
Refactor data manipulationin WorkspacePeoplePage

mountiny commented:
I see, the changes to `WorkspaceSidebar` must have been merged to main just before I have got up-to-date. However, I tried to merge main into this branch locally now to add the sidebar menu functionality and the `workspace/:policyID/people` route stopped working. I got redirected back to the chats and if I used the side menu and link to workspace and only get to the card page, but clicking _anywhere_ on the page brings me back to the chat page. 

Is there something I need to do to my testing account to be able to access the page with the latest changes merged? This should not be related to this PR.

marcaaron commented:
Hard to say from this end. Looks like you still have conflicts. Were they fixed correctly on your end? Maybe push up what you've got?

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * Remove user from the selectedEmployees list
+     *
+     * @param {Sting} login
```
`String`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * Add user from the selectedEmployees list
+     *
+     * @param {Sting} login
```
`String`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * Toggle user from the selectedEmployees list
+     *
+     * @param {Sting} login
```
`String`

marcaaron commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -0,0 +1,353 @@
+                        </View>
+                        <View style={[styles.w100, styles.mt4]}>
+                            {
+                                this.props.policy.employeeList && (
```
Not sure if this needs to be conditionally rendered anymore. There should be employees or `data` should always at least be an empty array now so I think we can just render the `FlatList`.

mountiny replied:
You are right, thanks!

mountiny committed:
Merge main into this branch

mountiny committed:
Remove the pressable area on fullscreen modal

mountiny committed:
Fix JSDoc typo and render FlatList unconditionally

mountiny commented:
@marcaaron Thank you for the reviews so far 🙏 I think I have narrowed down what was the problem with being unable to click on the Workspace settings sidebar menu. You cannot do it on `main` either. I had to add `isModal: false` to these `fullscreenModalScreenOptions`, because without this key, there is an overlay on the left side same as if it would be the right hand side settings modal and clicking on that part dismissed the fullscreen modal.

https://github.com/Expensify/Expensify.cash/blob/1a8b58ce36f44ea972382870ed404d5a813eb953/src/libs/Navigation/AppNavigator/AuthScreens.js#L215-L224

However, I still cannot access the `workspace/:policyId/people` route as I found out that `getStateFromPath` here is undefined. I am getting `Failed to parse the path to a navigation state.` error.
https://github.com/Expensify/Expensify.cash/blob/2852e12a1af44afd49dd9995d1fc8424fcf5b70f/src/libs/Navigation/linkTo.js#L13-L19

I was trying to find out why does the `/card` route work fine, but the `/people` route cannot be parsed, but I was not able to figure it out this morning. I was thinking you might know better as you wrote that logic. I will investigate more in the evening when you will be online, if this will stay a mystery by then.

mountiny committed:
Add spanish translations

mountiny committed:
Update finalized Spanish translations

marcaaron committed:
Fix route. Add back button press

marcaaron commented:
I pushed a change here that should fix the issue you were having. The trouble is that the linking config was not using an exact path but a relative one.

**parent path defined here:**
https://github.com/Expensify/Expensify.cash/blob/c5552954fd4ebcef0874edc10feaa80456f1de35/src/libs/Navigation/linkingConfig.js#L138-L139

**path for the card screen:** `:policyID/card`
**path for the people screen:** `workspace/:policyID/card`

so in this case react navigation couldn't find a correct screen since the people screen's final expected path is `workspace/workspace/:policyID/card`

marcaaron committed:
remove isModal

marcaaron committed:
Fix up click away handler issues

marcaaron reviewed (dismissed):
@mountiny I helped out with the click away handler parts. you were close, but the issue still persisted after the change.

@TomatoToaster maybe you can give this a once over + @shawnborton if you have any final design feedback. Here's how things are looking atm...

<img width="1440" alt="2021-06-30_05-57-17" src="https://user-images.githubusercontent.com/32969087/123995971-8ff6a980-d96a-11eb-946c-d2df3babc7ef.png">
<img width="1440" alt="2021-06-30_05-57-14" src="https://user-images.githubusercontent.com/32969087/123995978-91c06d00-d96a-11eb-879d-14ff7e0a935c.png">

mountiny commented:
@marcaaron Thank you very much for your help, time and patience with this PR. Really appreciated it 🙇‍♂️

marcaaron committed:
use white background

marcaaron commented:
>The right side of the workspace modal should have a white background but perhaps that is being addressed elsewhere by Rajat. 

Pushed a quick change for this. Rajat's changes are already merged so he's not working on it any longer.

>On that same note, I think we intended for the workspace modal to be inset a bit from the outer edges of the page (like the attachment modal) as you can see from this screenshot in Michelle's comment... again, I think perhaps that will be addressed elsewhere but just wanted to confirm.

I want to propose we explore this in a follow up. I looked into this briefly and couldn't find a quick solution. The attachment modal works a little differently from this modal since this one has the sidebar + react navigation to consider. I'm pretty confident we can get it done, but unsure how tricky it will be.

shawnborton commented:
Sounds good to me 👍

TomatoToaster reviewed (commented):
Code part LGTM from what I can see. When going to the settings page through e.cash it works fine, but I noticed one issue in testing and was wondering if anyone else could reproduce this.

I have a freePlan policy set up locally, so I tried to use the direct link to access the settings page (for me it was something like http://localhost:8080/workspace/4728CABFB50C430B/people). When I go to that direct link though the left hand side is blank for me instead of showing me `New Policy` ![image](https://user-images.githubusercontent.com/19364431/124017412-af1a2900-d9b4-11eb-9842-0c055719ef12.png)

This seems to happen if use any of my policyIDs not just the freePlan one. I was wondering if this happens to anyone else when they use a direct url. It might just be my `expensifail` account being wonky though.

marcaaron commented:
@TomatoToaster I can't reproduce this, but that seems like there is a problem for sure! Do you remember the order of steps you took to navigate there? Did you enter it into the browser url spot or click a link?

marcaaron commented:
Any errors in the console ?

marcaaron commented:
Oh wait. nvm this happens _very consistently_ for me haha

shawnborton commented:
Apologies for not catching this sooner but why are we splitting out the name from the email into two separate columns? Take a look at the mock here and we're just reusing the same pattern we use to display users everywhere else where we use a stacked pattern:
![image](https://user-images.githubusercontent.com/2319350/124019556-0cef4680-d99e-11eb-9f3f-bc8a511b7ead.png)

TomatoToaster commented:
To reproduce it I had just manually entered the Url with the policyID, I think you're seeing it now too 😄 . Though I'm only seeing this for an error:
```
Uncaught (in promise) DOMException: Failed to execute 'replaceState' on 'History': A history state object with URL 'http://workspace/4728CABFB50C430B/people' cannot be created in a document with origin 'http://localhost:8080' and URL 'http://localhost:8080//workspace/4728CABFB50C430B/people'.
    at Object.replace (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:33726:22)
    at onStateChange (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:34113:19)
    at callback (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:33881:13)
    at http://localhost:8080/app-25bda32969c003f0719a.bundle.js:26546:83
    at Array.forEach (<anonymous>)
    at Object.emit (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:26546:69)
    at http://localhost:8080/app-25bda32969c003f0719a.bundle.js:24488:13
    at invokePassiveEffectCreate (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:145329:20)
    at HTMLUnknownElement.callCallback (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:125792:14)
    at Object.invokeGuardedCallbackDev (http://localhost:8080/app-25bda32969c003f0719a.bundle.js:125841:16)
    ```
    
  I don't believe that error is related since I'm not able to catch it consistently 🤔

marcaaron committed:
get route params from whichever route we can. they might be out of order. previous logic only worked because there was a single route

marcaaron commented:
>Apologies for not catching this sooner but why are we splitting out the name from the email into two separate columns? Take a look at the mock here and we're just reusing the same pattern we use to display users everywhere else where we use a stacked pattern

Ah good catch. I wasn't quite paying attention either + the markup we used is not correct in that case. Going to clean this up a bit.

marcaaron committed:
clean up styles some more

marcaaron committed:
remove buttonRow

marcaaron committed:
clean up badge style

marcaaron commented:
Ok this is ready for another review.

@shawnborton I cleaned up the styles a bunch (and got rid of a bunch of markup we didn't need) so some of the padding might have shifted slightly, lemme know if it looks OK or if we should do some additional tweaks.

![2021-06-30_10-09-44](https://user-images.githubusercontent.com/32969087/124025723-81b98500-d98c-11eb-8b88-e6791e30d120.png)
![2021-06-30_10-09-29](https://user-images.githubusercontent.com/32969087/124025729-82eab200-d98c-11eb-9d2a-dc85c826d118.png)

shawnborton commented:
Looks great to me!

marcaaron committed:
remove sms domain

marcaaron committed:
Make entire row touchable

marcaaron committed:
fix bug discovered on android due to null policy

TomatoToaster commented:
Boom! It's working very nicely now. I am testing it with the banner here (right now it won't work on localhost on the master branch, but I'm updating it in this PR:https://github.com/Expensify/Web-Expensify/pull/31302  and will test it when that one goes live) 
![image](https://user-images.githubusercontent.com/19364431/124032398-87cc5780-d9c6-11eb-805c-8774a3816678.png)

![image](https://user-images.githubusercontent.com/19364431/124032385-826f0d00-d9c6-11eb-9b04-b74270fb8e61.png)

TomatoToaster commented on src/pages/home/sidebar/optionPropTypes.js:
```diff
@@ -22,7 +22,7 @@ const optionPropTypes = PropTypes.shape({
 
     // The array URLs of the person's avatar
-    icon: PropTypes.arrayOf(PropTypes.string),
+    icons: PropTypes.arrayOf(PropTypes.string),
```
Nice 👍🏾

shawnborton commented:
Another small design comment: maybe my eyes are playing tricks on me, but can someone confirm that the border width of the lines above & below the user rows are 1px? Same with the border divider between the left pane and right pane - is that 1px wide? Optically it feels like it might be 2 so just want to double check.

mountiny commented:
@shawnborton The divider between left and main part is actually 3px wide and the row divider are 2px wide. 

Should it all be only 1px?

shawnborton commented:
Yup - they should all be 1px, thanks!

mountiny commented:
Gonna update it!

mountiny committed:
Update border width to be 1px

mountiny committed:
Update Spanish translation

marcaaron commented:
Gonna merge this since it's blocking some other things + the only remaining blocker was the border which is fixed. Thanks for the reviews @TomatoToaster !!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

roryabraham commented:
Unfortunately the deploy comments are not working right now. This was deployed to staging yesterday.

kavimuru commented:
## Android - App is crashing after entered an email of an account to add to the workspace
## Expected Result: 
Able to add the number of accounts to the workspace
## Actual Result:
App is crashing after entered an email of account
## Actions Performed:
1. Lunch the app
2. Log in with expensifail account
3. Click on Plus button and Select Workspace
4. Tap the workspace
5. Tap the "People" option on the next screen
6. use the green Invite button which will bring in a modal from a side.
7. Enter an email of an account you want to add to the workspace
## Platform:
iOS 
mWeb 
Android ✔️
## Build:
1.0.76-0
## Notes/Images/Video:
![Bug5144843_Screenshot_20210708-220331_One_UI_Home](https://user-images.githubusercontent.com/43996225/125017598-7aa40e00-e041-11eb-8635-77a6b5e7f374.jpg)

https://user-images.githubusercontent.com/43996225/125017601-7b3ca480-e041-11eb-8c0a-3e9fd50ef4f9.mp4

[log file.txt](https://github.com/Expensify/Expensify.cash/files/6788419/log.file.txt)

Jag96 commented:
Looks like this is caused by [this line](https://github.com/Expensify/Expensify.cash/pull/3463/files#diff-5fd7c80f3a9b43595640753f4037ca2b7d0bebeca557a20c4a06f89dc5028ccfR106) because `policy.employeeList` is undefined for this flow. I'll move this to an issue and make a PR for this.

s77rt commented:
This PR caused a bug https://github.com/Expensify/App/issues/17166. The `WorkspacePeoplePage` suffers from stale data issue.
Selected employees are [stored in a state](https://github.com/Expensify/App/blob/3eb5818529b2de6ab4f7eabfbbf3e61b18def29a/src/pages/workspace/WorkspacePeoplePage.js#L62) that is filled with values from `policy.employeeList` prop but it does not react to changes to that prop e.g. if we clear `policy.employeeList` the `selectedEmployees` array would still hold values.
